### PR TITLE
[CORE] Move all columnar rules to post-columnar transitions

### PR DESCRIPTION
**Pull request by zhztheplayer**
This patch moves all columnar rules to post-columnar transition API. This works like we'll inject Gluten's custom rules to Spark's preparation rule list rather than in to columnar rules.

By doing this, we can treat all of Gluten's columnar rules as an individual Spark preparation rule that works on a complete Spark plan with C2Rs and R2Cs added . So we get full control of plan conversion rather than being clamped by Spark's rule `ApplyColumnarRulesAndInsertTransitions`. This will help on continuously improving Gluten's plan optimization capability and to make it not only limited to doing columnar conversions.

### CODE REVIEW

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 1]
```diff
@@ -520,39 +524,27 @@ private[extension] object ColumnarToRowLike {
...
-            child
-          }
+        case child if PlanUtil.isGlutenColumnarOp(child) =>
+          RowToColumnarExec(
+            BackendsApiManager.getSparkPlanExecApiInstance.genColumnarToRowExec(child))
+        case other => other
       })
-    case _ =>
-      plan.withNewChildren(plan.children.map(replaceWithVanillaRowToColumnar))
   }
```
Comment (zhztheplayer) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:548]: cc @ulysses-you

**Comment by zhztheplayer** [comment]
@rui-mo @PHILO-HE

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 2]
```diff
@@ -520,39 +524,27 @@ private[extension] object ColumnarToRowLike {
 }
 // This rule will try to add RowToColumnarExecBase and ColumnarToRowExec
 // to support vanilla columnar operators.
-case class VanillaColumnarPlanOverrides(session: SparkSession) extends Rule[SparkPlan] {
+case class InsertColumnarToColumnarTransitions(session: SparkSession) extends Rule[SparkPlan] {
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:527]: It seems the previous rule name is more readable. It is used to be compatible with vanilla columnar related things.

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 3]
```diff
@@ -684,29 +754,54 @@ case class ColumnarOverrideRules(session: SparkSession)
...
+   *
+   * See: https://github.com/oap-project/gluten/pull/4790
+   */
+  final override def preColumnarTransitions: Rule[SparkPlan] = plan => {
+    // To infer caller's property: ApplyColumnarRulesAndInsertTransitions#outputsColumnar.
+    OutputsColumnarTester.wrap(plan)
+  }
+
+  override def postColumnarTransitions: Rule[SparkPlan] = plan => {
+    val outputsColumnar = OutputsColumnarTester.inferOutputsColumnar(plan)
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:780]: why not use `plan.supportsColumnar` ?

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 4]
```diff
@@ -684,29 +754,54 @@ case class ColumnarOverrideRules(session: SparkSession)
...
+    val unwrapped = OutputsColumnarTester.unwrap(plan)
+    val vanillaPlan = ColumnarTransitions.insertTransitions(unwrapped, outputsColumnar)
+    withSuggestRules(suggestRules(outputsColumnar)).apply(vanillaPlan)
+  }
+
+  // Visible for testing.
+  def withSuggestRules(suggestRules: List[SparkSession => Rule[SparkPlan]]): Rule[SparkPlan] =
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:787]: How about naming `transformRules` ? The `suggest rules` is a bit hard to follow..

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala (hunk 2 above)
Comment (zhztheplayer) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:527]: Thanks for the explanation. 

Can we temporarily make the name describe what it actually does? If it's now converts from columnar to columnar, I will be inclined to use this new name.

I understand it could do more than columnar transitions but we can discuss on whether to add new logics to this rule or into other independent rules at that time.

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala (hunk 3 above)
Comment (zhztheplayer) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:780]: Spark has `supportsRowBased` since 3.3 . A plan node can be both `supportsColumnar == true` and `supportsRowBased == true`.  We should know about the caller's intention having which property requested.

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala (hunk 4 above)
Comment (zhztheplayer) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:787]: I'll be good to both. I'd change to `withTransformRules` if it's confusing.

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala (hunk 3 above)
Comment (ulysses-you) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:780]: If the plan support both columnar and row, then Spark should do thing. Does `plan.supportsColumnar` return different value with `OutputsColumnarTester.inferOutputsColumnar` ? It seems we never touch supportsRowBased.

Comment (zhztheplayer) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:780]: It's a rare case but it does covered by existing UTs...

Consider an `UnionExec` which supports both columnar and row-based execution, then even caller requested `outputsColumnar=false` through `ApplyColumnarRulesAndInsertTransitions#outputsColumnar`, we don't have a way to be aware of that since the `UnionExec` returns `true` for its `supportsColumnar` method.

So we should infer `outputsColumnar`'s exact value here rather than just calling `supportsColumnar`.

Comment (ulysses-you) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:780]: Can we move OutputsColumnarTester related code to a new file with the comment ? The ColumnarOverrides file is too big.

**Review by JkSelf — APPROVED** [review]
Great work. LGTM. Thanks.

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala (hunk 3 above)
Comment (zhztheplayer) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:780]: I was planning on splitting the file too. Let's do that in another patch. Thanks for the suggestion.
